### PR TITLE
Speedup r_anal_type_func_guess

### LINKED ISSUE #8028: r_anal_type_func_guess is very slow
r_anal_type_func_guess is currently using O(n^3)  algorithm for checking each substring  where n is length of string. This is very slow for c++ executables with a lot of long names. Checking all the substrings also causes plenty of false positives like pthread_something=read and operand_something=rand.

Checking only prefixes and suffixes reduce `aa` time for cmake from 30 to 3 minutes.  Only test broken by checking only prefixes and suffixes was "type-matching/Function definition (autorename)" because "sub.KERNEL32.dll_ExitProcess_850" has KERNEL32.dll_ and _850 added. I assume that both of them are added by radare. A long term solutions would be to properly distinguish between different type of data instead of throwing everything into single string. Simpler solution just to get anal_type_func_guess fast would be to try detecting module name and remove it. 

I guess that currently there aren't too many test cases checking quality of analysis so there might be other cases that have both some prefix and suffix added around function name. Any ideas of what could those be?

### PR DESCRIPTION
**Pull request by karliss**
Fixes #8028.

Speed improvements aren't as impressive as mentioned in #8028 44min->9.5min.  Seeing that both times have increased I guess that some other slow code was recently added.

### CODE REVIEW

File: libr/anal/types.c [hunk 1]
```diff
@@ -299,15 +299,29 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
 
 #define MIN_MATCH_LEN 4
 
-// TODO: Future optimizations
-// - symbol names are long and noisy, reduce searchspace for match here
-//   by preprocessing to delete noise, as much as we can.
-// - We ignore all func.*, but thats maybe most of sdb entries
-//   could maybe eliminate this work (strcmp).
+static char *type_func_try_guess(RAnal *anal, char *name) {
+	if (strlen (name) < MIN_MATCH_LEN) {
```
Comment (radare) [inline libr/anal/types.c]: use r_str_nlen, will be faster for strings longer than MIN_MATCH_LEN

File: libr/anal/types.c [hunk 2]
```diff
@@ -299,15 +299,29 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
...
+		return NULL;
+	}
+	if (sdb_exists (anal->sdb_types, name)) {
+		const char *res = sdb_const_get (anal->sdb_types, name, 0);
```
Comment (radare) [inline libr/anal/types.c]: if sdb_const_get returns NULL, you can assume that sdb_exists, so you can save 1 sdb call here.

File: libr/anal/types.c [hunk 3]
```diff
@@ -299,15 +299,29 @@ R_API char *r_anal_type_func_args_name(RAnal *anal, const char *func_name, int i
...
+		bool is_func = res && !strcmp ("func", res);
+		if (is_func) {
+			char *ret = strdup (name);
```
Comment (radare) [inline libr/anal/types.c]: return strdup (name);

File: libr/anal/types.c [hunk 4]
```diff
@@ -317,34 +331,50 @@ R_API char *r_anal_type_func_guess(RAnal *anal, char *func_name) {
...
+	if ((result = type_func_try_guess (anal, str))) {
+		return result;
+	}
+	str = strdup (str);
+	// some names are in format module.dll_function_number, try to remove those
+	// also try module.dll_function and function_number
+	if ((first = strchr (str, '_'))) {
+		last = first;
+		j = first - str;
+		while (str[++j]) {
```
Comment (radare) [inline libr/anal/types.c]: this loop looks like last = r_str_lchr (str, '_')

**Comment by radare** [comment]
comparing to which commit?

File: libr/anal/types.c (hunk 1 above)
Comment (karliss) [inline libr/anal/types.c]: Done.

File: libr/anal/types.c (hunk 2 above)
Comment (karliss) [inline libr/anal/types.c]: I was considering it that when I read that code first time. Done.

File: libr/anal/types.c (hunk 3 above)
Comment (karliss) [inline libr/anal/types.c]: Done.

File: libr/anal/types.c (hunk 4 above)
Comment (karliss) [inline libr/anal/types.c]: Done.

**Comment by karliss** [comment]
It seems that time difference 30->3min vs 44->9min wasn't caused by other slow commit. I was running test    on two different cmake copies.

**Comment by radare** [comment]
awesome! will merge after travis. thanks!

**Pull request merged**